src/vm/runners/builtin_runner/keccak.rs

```diff
@@ -0,0 +1,610 @@
+                base.try_into()
+                    .map_err(|_| MemoryError::AddressInTemporarySegment(base))?,
+            )
+            .ok_or(MemoryError::MissingSegmentUsedSizes)
+    }
+
+    pub fn get_used_cells_and_allocated_size(
+        &self,
+        vm: &VirtualMachine,
+    ) -> Result<(usize, usize), MemoryError> {
+        let ratio = self.ratio as usize;
+        let cells_per_instance = self.cells_per_instance;
+        let min_step = ratio * self.instances_per_component as usize;
+        if vm.current_step < min_step {
+            Err(MemoryError::InsufficientAllocatedCells)
+        } else {
+            let used = self.get_used_cells(vm)?;
+            let size = cells_per_instance as usize
+                * safe_div_usize(vm.current_step, ratio)
+                    .map_err(|_| MemoryError::InsufficientAllocatedCells)?;
+            Ok((used, size))
+        }
+    }
+
+    pub fn get_used_instances(&self, vm: &VirtualMachine) -> Result<usize, MemoryError> {
+        let used_cells = self.get_used_cells(vm)?;
+        Ok(div_ceil(used_cells, self.cells_per_instance as usize))
+    }
+
+    pub fn final_stack(
+        &self,
+        vm: &VirtualMachine,
+        pointer: Relocatable,
+    ) -> Result<(Relocatable, usize), RunnerError> {
+        if self._included {
+            if let Ok(stop_pointer) = vm
+                .get_relocatable(&(pointer.sub(1)).map_err(|_| RunnerError::FinalStack)?)
+                .as_deref()
+            {
+                if self.base() != stop_pointer.segment_index {
+                    return Err(RunnerError::InvalidStopPointer("keccak".to_string()));
+                }
+                let stop_ptr = stop_pointer.offset;
+                let num_instances = self
+                    .get_used_instances(vm)
+                    .map_err(|_| RunnerError::FinalStack)?;
+                let used_cells = num_instances * self.cells_per_instance as usize;
+                if stop_ptr != used_cells {
+                    return Err(RunnerError::InvalidStopPointer("keccak".to_string()));
+                }
+
+                Ok((
+                    pointer.sub(1).map_err(|_| RunnerError::FinalStack)?,
+                    stop_ptr,
+                ))
+            } else {
+                Err(RunnerError::FinalStack)
+            }
+        } else {
+            let stop_ptr = self.base() as usize;
+            Ok((pointer, stop_ptr))
+        }
+    }
+
+    pub fn get_memory_accesses(
+        &self,
+        vm: &VirtualMachine,
+    ) -> Result<Vec<Relocatable>, MemoryError> {
+        let base = self.base();
+        let segment_size = vm
+            .segments
+            .get_segment_size(
+                base.try_into()
+                    .map_err(|_| MemoryError::AddressInTemporarySegment(base))?,
+            )
+            .ok_or(MemoryError::MissingSegmentUsedSizes)?;
+
+        Ok((0..segment_size).map(|i| (base, i).into()).collect())
+    }
+}
+
```

Comment:
The implementation of `get_used_diluted_check_units `is missing here, it should also me added to the method  `get_used_diluted_check_units` on the BuiltinRunner enum definition (mod.rs)